### PR TITLE
mnist_tfrecord.py added

### PR DESCRIPTION
**Pull request by ahundt**
This PR implements an mnist run via a TFRecord dataset.

This PR requires https://github.com/fchollet/keras/pull/7060 to be merged first.

Also see for details:

- #6928 yield_ops with `labels` param
- #7075/#7046 internal fixes to `Model()`

**Update:** Performance bug Fixed by defaulting `compile(loss=None)`

It seems with the current implementation, the **external loss version suffers from 20x performance degradation** compared to `mnist_tfrecord.py` in #6928 which has a more direct C++ backend pipeline.

- **internal_fixes_tfrecord #7075/#7046**
  - Test accuracy: 0.795400005609
  - Training time: 23:17.18 (m:s) total
  - `mnist_tfrecord.py` with external loss

- **yield_ops #6928**
  - Test accuracy: 0.967999995768
  - Training time: 1:22.24 (m:s) total
  - `mnist_tfrecord.py` with internal loss and no `K.placeholder()` objects

- **mnist_cnn.py**
  - Test accuracy: 0.9902
  - Training time: 1:13.12 (m:s) total

Note that the slightly lower performance of mnist_tfrecord.py #6928 is expected, because it is a pipeline designed for quickly loading large datasets that don't fit in memory, and writing the records and graph setup takes several extra seconds.

other relevant PRs: #7061, #7072, #6928, #7046

### CODE REVIEW

**Comment by fchollet** [comment]
> It seems with the current implementation, the external loss version suffers from 20x performance degradation compared to mnist_tfrecord.py in #6928 which has a more direct C++ backend pipeline.

Unless you have an implementation issue, the underlying TF graphs will be identical and will thus perform the same.

Can you please summarize the extent and goals of the changes you are introducing in this new string of PRs? It is very difficult for me to follow what you are trying to do and how, and why it is necessary.

**Comment by fchollet** [comment]
I think in general we need an official "API design review" process, where any new feature or API change would:
- Have their API choices discussed and approved beforehand
- Have their high-level technical choices discussed and approved beforehand

So that PRs would be limited to correctness and style reviews, and would be agnostic to higher-level decision making. Separating API concerns and architecture concerns from code/implementation concerns.

With that setup, you would submit a design doc, we would discuss it and come up with a decision, and only then would you prepare PRs. There would be no more questions along the lines of "what is the purpose of this PR".

I will work on setting up such a process.

**Comment by fchollet** [comment]
I have added an API design process to CONTRIBUTING.md: https://github.com/fchollet/keras/blob/master/CONTRIBUTING.md

Here's the design doc template: https://docs.google.com/document/d/1ZXNfce77LDW9tFAj6U5ctaJmI5mT7CQXOFMEAZo-mAA/edit#

Again, the goal is to separate design concerns from implementation concerns.

**Comment by fchollet** [comment]
> There is a key difference, master is guaranteed to create a placeholder for every output while #6928 directly feeds the RecordInput tensors. That's the most likely source of the real performance discrepancy.

It's optional. Unless you intend on passing Numpy data to an output, you don't have to create a placeholder. If your loss is external, you wouldn't create one.

> Update: I just remembered that depends if skip_indices is None or not, but I think they should not be None because of add_loss... hmm that goes into topology.py... this model.compile() code path could use some more cleanup. I'm not sure yet. :-)

It isn't None because of `add_loss`, it is None if you set the loss of the corresponding output to None in `compile`. Which is what you should do if you don't intend on passing Numpy data to that output (i.e. if your loss is external).

**Comment by ahundt** [comment]
great, that worked! Just a few seconds slower now. Perhaps it was a bad merge, I had created so many branches.

```
Test accuracy: 0.957299995482
python mnist_tfrecord.py  219.07s user 51.71s system 313% cpu 1:26.48 total
```

**Comment by ahundt** [comment]
I do think some of the doc proposals improve the delightfulness of the API by avoiding this external loss workaround, so performance will be ideal even with the accidental corssentropy.

**Comment by ahundt** [comment]
closing in favor of https://github.com/fchollet/keras/pull/7113
